lib/schema/generateSchema.js

```diff
@@ -4,46 +4,53 @@ var fs = require('fs'),
 
 exports.generateSchema = function generateSchema(next) {
 	var self = this;
-	if (self.config.schemaRefresh === false) {
-		self.logger.debug('skipping generation of schema, config.schemaRefresh is false');
-		return next && next();
-	}
-
-	/*
-	 Use a hashed filename based on the configured authentication so if
-	 they change their config, we will update the schema too.
-	 */
 	var confDir = self.getCacheDir(),
 		hashedFilename = 'salesforce-schema-' + self.md5((self.config.generateModels || []).join('.') + self.config.username + self.config.token) + '.json';
 
-	// If schemaCache is false, we look it up every time.
-	if (self.config.schemaCache === false) {
-		self.cleanGeneratedModels();
-		return self.describeGlobal(confDir, hashedFilename, next);
-	}
-
 	var cachedFileName = path.join(confDir, hashedFilename),
 		schemaRefresh = self.config.schemaRefresh === undefined ? 3.6e+6 : self.config.schemaRefresh;
 
-	fs.exists(cachedFileName, function schemaExists(exists) {
-		if (!exists) {
-			self.logger.debug('schema file not found, will generate');
-			self._cachedSchema = null;
-			self.cleanGeneratedModels();
-			self.describeGlobal(confDir, hashedFilename, next);
+	if (shouldGenerateNewSchema()) {
+		cleanModelsAndGenerateNewSchema();
+	}
+	else {
+		if (self.config.schemaRefresh === false) {
+			useCachedVersion();
 		}
 		else {
 			checkCachedVersion();
 		}
-	});
+	}
+
+	function shouldGenerateNewSchema() {
+		return (self.config.schemaCache === false) || !(fs.existsSync(cachedFileName));
+	}
+
+	function cleanModelsAndGenerateNewSchema() {
+		self._cachedSchema = null;
+		self.cleanGeneratedModels();
+		self.describeGlobal(confDir, hashedFilename, next);
```

Comment:
Pay attention to cachedFileName variable in this file. Then check  why we need to pass down both confDir and hashedFilename. Is it possible to do more optimizations here?